### PR TITLE
Add dsa_wycheproof_test

### PR DESCRIPTION
**Pull request by karel-m**

### CODE REVIEW

**Comment by karel-m** [comment]
In the latest release of my libtomcrypt perl bindings (based on the latest release/1.18.0 branch) I turned on wycheproof tests (for RSA, DSA, ECDH) which mostly pass except the test `type=DSAVer/SHA224 tcId=55 comment='appending unused 0's' expected-result=invalid` e.g.
* http://www.cpantesters.org/cpan/report/77dce23e-9a70-11e7-af1e-236b50cc459e
* http://www.cpantesters.org/cpan/report/e6371e2e-9a21-11e7-8929-daabf0ac1f1e
* both happened on ARM architecture

So I have decided to add this "sometimes failing" test to libtomcrypt's `dsa_test()`. It passes fine on my laptop; however, travis-ci is not quite happy about it.

**Comment by karel-m** [comment]
Update: the troubles with Travis-CI turned out to be TFM related. It seems that `radix_to_bin` does not work correctly with TFM - when passed a hexadecimal string of 1676 characters it returns `CRYPT_OK` but the binary output is only 576 bytes (instead of expected 838).

Anyway I am for merging this extra DSA test and see whether somebody will report a failure on ARM or not.

**Comment by sjaeckel** [comment]
looks like this happened with gcc and clang...

/cc @sebastianas

**Comment by karel-m** [comment]
Those failures happened on ARM-32bit-LE. Can anybody checkout `pr/dsa-test-wycheproof` and build & run `./test` on this architecture?

**Comment by ksherlock** [comment]
```
$ uname -a
Linux qnap 3.2.0-4-orion5x #1 Debian 3.2.84-1 armv5tel GNU/Linux

$ git branch
  develop
* pr/dsa-test-wycheproof

$ ./test
build == git-v1.18.0-rc3-48-ge2b6631
LibTomCrypt 1.18.0-rc3 (www.libtom.net)
LibTomCrypt is public domain software.


Endianness: little (32-bit words)
Clean stack: disabled
Ciphers built-in:
   Blowfish
   RC2
   RC5
   RC6
   Safer+
   Safer
   Rijndael
   XTEA
   Twofish (tables)
   DES
   CAST5
   Noekeon
   Skipjack
   Khazad
   Anubis  (tweaked)
   KSEED
   KASUMI
   MULTI2
   Camellia
Stream ciphers built-in:
   ChaCha
   RC4
   SOBER128

Hashes built-in:
   SHA3
   SHA-512
   SHA-384
   SHA-512/256
   SHA-256
   SHA-512/224
   SHA-224
   TIGER
   SHA1
   MD5
   MD4
   MD2
   RIPEMD128
   RIPEMD160
   RIPEMD256
   RIPEMD320
   WHIRLPOOL
   BLAKE2S
   BLAKE2B
   CHC_HASH

Block Chaining Modes:
   CFB
   OFB
   ECB
   CBC
   CTR
   LRW (tables) 
   F8
   XTS

MACs:
   HMAC
   OMAC
   PMAC
   PELICAN
   XCBC
   F9
   POLY1305
   BLAKE2S MAC
   BLAKE2B MAC

ENC + AUTH modes:
   EAX
   OCB
   OCB3
   CCM
   GCM (tables) 
   CHACHA20POLY1305

PRNG:
   Yarrow (LTC_YARROW_AES=2)
   SPRNG
   RC4
   ChaCha20
   Fortuna (LTC_FORTUNA_POOLS=32, LTC_FORTUNA_WD=10)
   SOBER128

PK Algs:
   RSA (with blinding and CRT hardening)
   DH
   ECC (with blinding)
   DSA

MPI (Math):
   LTC_MPI
   LTC_MILLER_RABIN_REPS=35

Compiler:
   GCC compiler 7.2.0.

Various others:  ARGTYPE=0  ADLER32  BASE64  BASE64-URL-SAFE  CRC32  DER  LTC_DER_MAX_PUBKEY_SIZE=4096  PKCS#1  PKCS#5  HKDF  LTC_DEVRANDOM  LTC_TRY_URANDOM_FIRST  LTC_RNG_GET_BYTES  LTC_RNG_MAKE_PRNG  LTC_HASH_HELPERS  LTC_TEST  LTC_FILE_READ_BUFSIZE=8192  LTC_ECC_SHAMIR 

math provider = libtommath
MP_DIGIT_BIT = 28

store_test..........passed      0.037ms
rotate_test.........passed      0.038ms
misc_test...........passed      1.920ms
mpi_test............passed     26.823ms
cipher_hash_test....passed    424.496ms
mac_test............passed    220.373ms
modes_test..........passed     42.537ms
der_test............passed  59097.761ms
pkcs_1_test.........passed     44.507ms
pkcs_1_pss_test.....passed    630.505ms
pkcs_1_oaep_test....passed    627.422ms
pkcs_1_emsa_test....passed    829.914ms
pkcs_1_eme_test.....passed    828.633ms
rsa_test............passed  67807.963ms
dh_test.............passed 990140.469ms
ecc_tests...........passed 721116.053ms
dsa_test............passed  15634.199ms
katja_test..........nop
file_test...........passed      3.515ms
multi_test..........passed      1.390ms
prng_test...........passed     53.802ms

SUCCESS: passed=20 failed=0 nop=1 duration=1857.6sec real=1857.5sec

**Comment by sjaeckel** [comment]
And with TFM?

**Comment by karel-m** [comment]
> SUCCESS: passed=20 failed=0

Thanks, although a failure would have been more valuable in this case 😃 

Ad TFM: the failures from my perl bindings are from LTC+LTM; the trouble with `radix_to_bin()` was "fixed" by switching from hex to Base64. In other words TFM should be fine.

**Comment by sjaeckel** [comment]
mh okay

1. do you still have the failing testcase of TFM?
2. what @ksherlock tested was armv5, we need a test on armv6 as used in your tests

**Comment by karel-m** [comment]
> do you still have the failing testcase of TFM?

#294

> we need a test on armv6 as used in your tests

I have installed `Raspbian GNU/Linux 9.1 (stretch)` in QEMU:
```
$ uname -a
Linux raspberrypi 4.4.13+ #12 Thu Sep 15 12:00:26 CEST 2016 armv6l GNU/Linux
```
but no failure (neither with this PR nor with my perl bindings).

I'll try to contact the guys running those perl cpan smokers and ask for more details.

**Comment by karel-m** [comment]
More info about perl bindings failure
```
# Testing CryptX 0.053_001, Perl 5.026000 linux /srv/smoker/perl5/perlbrew/perls/perl-5.26.0/bin/perl
# osname       = linux
# myarchname   = armv6l-linux
# myuname      = linux smoker 3.18.0-trunk-rpi #1 preempt raspbian 3.18.5-1~exp1+rpi19+stretch (2015-10-14) armv6l gnulinux 
# intsize      = 4
# longsize     = 4
# longlongsize = 8
# ptrsize      = 4
# byteorder    = 1234
# 
# LibTomCrypt 1.18.0-rc2 (www.libtom.net)
# LibTomCrypt is public domain software.
# 
# 
# Endianness: little (32-bit words)
# Clean stack: disabled
# Ciphers built-in:
#    Blowfish
#    RC2
#    RC5
#    RC6
#    Safer+
#    Safer
#    Rijndael
#    XTEA
#    Twofish (tables)
#    DES
#    CAST5
#    Noekeon
#    Skipjack
#    Khazad
#    Anubis  (tweaked)
#    KSEED
#    KASUMI
#    MULTI2
#    Camellia
# Stream ciphers built-in:
#    ChaCha
#    RC4
#    SOBER128
# 
# Hashes built-in:
#    SHA3
#    SHA-512
#    SHA-384
#    SHA-512/256
#    SHA-256
#    SHA-512/224
#    SHA-224
#    TIGER
#    SHA1
#    MD5
#    MD4
#    MD2
#    RIPEMD128
#    RIPEMD160
#    RIPEMD256
#    RIPEMD320
#    WHIRLPOOL
#    BLAKE2S
#    BLAKE2B
#    CHC_HASH
# 
# Block Chaining Modes:
#    CFB
#    OFB
#    ECB
#    CBC
#    CTR
#    LRW (tables) 
#    F8
#    XTS
# 
# MACs:
#    HMAC
#    OMAC
#    PMAC
#    PELICAN
#    XCBC
#    F9
#    POLY1305
#    BLAKE2S MAC
#    BLAKE2B MAC
# 
# ENC + AUTH modes:
#    EAX
#    OCB
#    OCB3
#    CCM
#    GCM (tables) 
#    CHACHA20POLY1305
# 
# PRNG:
#    Yarrow (LTC_YARROW_AES=2)
#    SPRNG
#    RC4
#    ChaCha20
#    Fortuna (LTC_FORTUNA_POOLS=32, LTC_FORTUNA_WD=10)
#    SOBER128
# 
# PK Algs:
#    RSA (with blinding and CRT hardening)
#    DH
#    ECC (with blinding)
#    DSA
# 
# MPI (Math):
#    LTC_MPI
#    LTM_DESC
#    LTC_MILLER_RABIN_REPS=35
# 
# Compiler:
#    GCC compiler 7.2.0.
# 
# Various others:  ARGTYPE=0  ADLER32  BASE64  BASE64-URL-SAFE  CRC32  DER  LTC_DER_MAX_PUBKEY_SIZE=4096  PKCS#1  PKCS#5  HKDF  LTC_DEVRANDOM  LTC_TRY_URANDOM_FIRST  LTC_RNG_GET_BYTES  LTC_RNG_MAKE_PRNG  LTC_HASH_HELPERS  LTC_FILE_READ_BUFSIZE=8192  LTC_NO_TEST  LTC_ECC_SHAMIR 
# 
# MP_PROVIDER  = LibTomMath
# MP_DIGIT_BIT = 28
# 
```

It was Raspberry Pi 1B, running Raspbian and a perl-brewed perl 5.26.

**Comment by karel-m** [comment]
@ksherlock I have update `pr/dsa-test-wycheproof` - could you please pull the latest version and run it again?

**Comment by ksherlock** [comment]
```
$ git branch -v
  develop                e1fcd3a add hashsum multi-hash option
* pr/dsa-test-wycheproof e9a9d61 sha256 was wrong, use sha224

$ ./test
build == git-v1.18.0-rc3-48-ge2b6631
LibTomCrypt 1.18.0-rc3 (www.libtom.net)
LibTomCrypt is public domain software.


Endianness: little (32-bit words)
Clean stack: disabled
Ciphers built-in:
   Blowfish
   RC2
   RC5
   RC6
   Safer+
   Safer
   Rijndael
   XTEA
   Twofish (tables)
   DES
   CAST5
   Noekeon
   Skipjack
   Khazad
   Anubis  (tweaked)
   KSEED
   KASUMI
   MULTI2
   Camellia
Stream ciphers built-in:
   ChaCha
   RC4
   SOBER128

Hashes built-in:
   SHA3
   SHA-512
   SHA-384
   SHA-512/256
   SHA-256
   SHA-512/224
   SHA-224
   TIGER
   SHA1
   MD5
   MD4
   MD2
   RIPEMD128
   RIPEMD160
   RIPEMD256
   RIPEMD320
   WHIRLPOOL
   BLAKE2S
   BLAKE2B
   CHC_HASH

Block Chaining Modes:
   CFB
   OFB
   ECB
   CBC
   CTR
   LRW (tables) 
   F8
   XTS

MACs:
   HMAC
   OMAC
   PMAC
   PELICAN
   XCBC
   F9
   POLY1305
   BLAKE2S MAC
   BLAKE2B MAC

ENC + AUTH modes:
   EAX
   OCB
   OCB3
   CCM
   GCM (tables) 
   CHACHA20POLY1305

PRNG:
   Yarrow (LTC_YARROW_AES=2)
   SPRNG
   RC4
   ChaCha20
   Fortuna (LTC_FORTUNA_POOLS=32, LTC_FORTUNA_WD=10)
   SOBER128

PK Algs:
   RSA (with blinding and CRT hardening)
   DH
   ECC (with blinding)
   DSA

MPI (Math):
   LTC_MPI
   LTC_MILLER_RABIN_REPS=35

Compiler:
   GCC compiler 7.2.0.

Various others:  ARGTYPE=0  ADLER32  BASE64  BASE64-URL-SAFE  CRC32  DER  LTC_DER_MAX_PUBKEY_SIZE=4096  PKCS#1  PKCS#5  HKDF  LTC_DEVRANDOM  LTC_TRY_URANDOM_FIRST  LTC_RNG_GET_BYTES  LTC_RNG_MAKE_PRNG  LTC_HASH_HELPERS  LTC_TEST  LTC_FILE_READ_BUFSIZE=8192  LTC_ECC_SHAMIR 

math provider = libtommath
MP_DIGIT_BIT = 28

store_test..........passed      0.040ms
rotate_test.........passed      0.038ms
misc_test...........passed      1.900ms
mpi_test............passed     21.129ms
cipher_hash_test....passed    430.671ms
mac_test............passed    221.902ms
modes_test..........passed     42.324ms
der_test............passed  59094.050ms
pkcs_1_test.........passed     43.177ms
pkcs_1_pss_test.....passed    627.941ms
pkcs_1_oaep_test....passed    628.664ms
pkcs_1_emsa_test....passed    864.495ms
pkcs_1_eme_test.....passed    830.470ms
rsa_test............passed  63525.862ms
dh_test.............passed 997150.779ms
ecc_tests...........passed 695970.022ms
dsa_test............passed  26096.888ms
katja_test..........nop
file_test...........passed      3.610ms
multi_test..........passed      1.791ms
prng_test...........passed     52.794ms

SUCCESS: passed=20 failed=0 nop=1 duration=1845.7sec real=1845.6sec

**Comment by karel-m** [comment]
TL;DR fixed by 155e29d - let's see what's Travis' opinion.
